vector/src/main/java/im/vector/riotx/features/terms/ReviewTermsViewModel.kt

```diff
@@ -0,0 +1,157 @@
+ */
+package im.vector.riotx.features.terms
+
+import com.airbnb.mvrx.ActivityViewModelContext
+import com.airbnb.mvrx.Loading
+import com.airbnb.mvrx.MvRxViewModelFactory
+import com.airbnb.mvrx.Success
+import com.airbnb.mvrx.Uninitialized
+import com.airbnb.mvrx.ViewModelContext
+import com.squareup.inject.assisted.Assisted
+import com.squareup.inject.assisted.AssistedInject
+import im.vector.matrix.android.api.MatrixCallback
+import im.vector.matrix.android.api.session.Session
+import im.vector.matrix.android.api.session.terms.GetTermsResponse
+import im.vector.riotx.core.extensions.exhaustive
+import im.vector.riotx.core.platform.VectorViewModel
+import timber.log.Timber
+
+class ReviewTermsViewModel @AssistedInject constructor(
+        @Assisted initialState: ReviewTermsViewState,
+        private val session: Session
+) : VectorViewModel<ReviewTermsViewState, ReviewTermsAction, ReviewTermsViewEvents>(initialState) {
+
+    @AssistedInject.Factory
+    interface Factory {
+        fun create(initialState: ReviewTermsViewState): ReviewTermsViewModel
+    }
+
+    companion object : MvRxViewModelFactory<ReviewTermsViewModel, ReviewTermsViewState> {
+
+        @JvmStatic
+        override fun create(viewModelContext: ViewModelContext, state: ReviewTermsViewState): ReviewTermsViewModel? {
+            val activity: ReviewTermsActivity = (viewModelContext as ActivityViewModelContext).activity()
+            return activity.viewModelFactory.create(state)
+        }
+    }
+
+    lateinit var termsArgs: ServiceTermsArgs
+
+    override fun handle(action: ReviewTermsAction) {
+        when (action) {
+            is ReviewTermsAction.LoadTerms          -> loadTerms(action)
+            is ReviewTermsAction.MarkTermAsAccepted -> markTermAsAccepted(action)
+            ReviewTermsAction.Accept                -> acceptTerms()
+        }.exhaustive
+    }
+
+    private fun markTermAsAccepted(action: ReviewTermsAction.MarkTermAsAccepted) = withState { state ->
+        val newList = state.termsList.invoke()?.map {
+            if (it.url == action.url) {
+                it.copy(accepted = action.accepted)
+            } else {
+                it
+            }
+        }
+
+        if (newList != null) {
+            setState {
+                state.copy(
+                        termsList = Success(newList)
+                )
+            }
+        }
+    }
+
+    private fun acceptTerms() = withState { state ->
+        val acceptedTerms = state.termsList.invoke() ?: return@withState
+
+        if (acceptedTerms.any { it.accepted.not() }) {
+            // Should not happen
+            _viewEvents.post(ReviewTermsViewEvents.Failure(IllegalStateException("Please accept all terms"), false))
+            return@withState
+        }
+
+        setState {
+            copy(
+                    acceptingTerms = Loading()
+            )
+        }
+
+        val agreedUrls = acceptedTerms.map { it.url }
+
+        session.agreeToTerms(
```

Comment:
viewModelScope.launch ?

Reply:
Same question

Reply:
Yes, but we should perform this change globally in the app, right?